feat(sdk-coin-avaxp): add avaxp new SDK

noel-bitgo opened the pull request:
CR-1073

TICKET: CR-1073

<!--
# Please be aware of the following when making your pull request:

## Description

Please include a summary of your proposed changes and which issue is being addressed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Issue Number

Internal Users - Please include the related internal tracking number (e.g. BG-000000).
External Users - Please link to any relevant github issues as necessary.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] My code compiles correctly for both Node and Browser environments
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My commits follow [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) and I have properly described any BREAKING CHANGES
- [ ] The ticket or github issue was included in the commit message as a reference
- [ ] I have made corresponding changes to the documentation and on any new/updated functions and/or methods - [jsdoc](https://jsdoc.app/)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
-->

sol-mocha reviewed (changes requested):
https://github.com/BitGo/BitGoJS/pull/4119

zahin-mohammad reviewed (changes requested):
Few comments:
- Browser tests won't pass due to cypress
- Changing min version to node 18 is a breaking change, you need to add the appropriate footer tag for this

I already did both in my node 16 PR, please rebate off of mine.
https://github.com/BitGo/BitGoJS/pull/4119

noel-bitgo committed:
feat(sdk-coin-avaxp): add new avaxp library

new AVAX library minimum node version is node 20.11.1
https://github.com/ava-labs/avalanchejs

this repo supports both node 18 and node 20

this PR uses a forked version of the avax library that supports node 18
https://www.npmjs.com/package/bitgo-aaron-avalanchejs?activeTab=readme

TICKET: CR-1073

CapnMigraine committed:
feat(sdk-coin-avaxp): add skeleton for addPermissionLessValidator tx

TICKET: CR-1073

noel-bitgo commented on modules/sdk-coin-avaxp/src/lib/permissionlessValidatorTxBuilder.ts:
```diff
@@ -0,0 +1,173 @@
+   *
+   * @param blsKey
+   */
+  blsKey(blsKey: string): this {
```
oh i used `blsPublicKey`

noel-bitgo commented on modules/sdk-coin-avaxp/src/lib/permissionlessValidatorTxBuilder.ts:
```diff
@@ -0,0 +1,173 @@
+import { AddPermissionlessValidatorTx } from 'bitgo-aaron-avalanchejs/dist/serializable/pvm';
+import { Tx } from './iface';
+
+export class PermissionlessValidatorTxBuilder {
```
add [getPermissionlessValidatorTxBuilder](https://github.com/BitGo/BitGoJS/blob/e441e877057c8ba4f22d1b721423c0cbebc7bb3b/modules/sdk-coin-avaxp/src/lib/transactionBuilderFactory.ts#L77)

zahin-mohammad commented on modules/sdk-coin-avaxp/package.json:
```diff
@@ -51,6 +51,7 @@
     "@bitgo/sdk-core": "^26.3.0",
     "@bitgo/statics": "^48.1.0",
     "@bitgo/utxo-lib": "^9.34.0",
+    "bitgo-aaron-avalanchejs": "^4.0.5",
```
If we know this is a working fix, can we fork this repo into a bitgo controlled repo?

noel-bitgo replied:
yes on it

andrew-scott-fischer commented on modules/sdk-coin-avaxp/package.json:
```diff
@@ -51,6 +51,7 @@
     "@bitgo/sdk-core": "^26.3.0",
     "@bitgo/statics": "^48.1.0",
     "@bitgo/utxo-lib": "^9.34.0",
+    "bitgo-aaron-avalanchejs": "^4.0.5",
```
This needs to be 
```suggestion
    "@bitgo/avalanchejs": "1.0.0-alpha.3",
```

DinshawKothari replied:
Thanks @andrew-scott-fischer. Will use this in PR - https://github.com/BitGo/BitGoJS/pull/4366 too.

andrew-scott-fischer committed:
chore: update BitGoJS to use `typescript@4.7.4`

Ticket: VL-1844

andrew-scott-fischer committed:
fix: ignore type-checks of error objects

In TypeScript 4.4 error objects changed from `any` to `unknown`.
Rather than fix this, this change ignores it.

Ticket: VL-1844

andrew-scott-fischer committed:
fix: cast as `unknown` before recasting

Exising code cast one object as another; the type might be wrong
but I'm not a SME. Casting `as unknown` to suppress TypeScript
errors.

Ticket: VL-1844

andrew-scott-fischer commented:
> > Few comments:
> > 
> > * Browser tests won't pass due to cypress

Looks like `browser-test`s did pass? Why wasn't that expected? @zahin-mohammad @noel-bitgo ?

zahin-mohammad commented:
> > > Few comments:
> > > 
> > > * Browser tests won't pass due to cypress
> 
> Looks like `browser-test`s [did pass](https://github.com/BitGo/BitGoJS/actions/runs/8440139442/job/23116363822?pr=4361)? Why wasn't that expected? @zahin-mohammad @noel-bitgo ?

Honestly not 100% sure why, but upgrading cypress fixed it for me. https://bitgo.slack.com/archives/C010AEXLLCR/p1711229931703669

andrew-scott-fischer commented:
> * Changing min version to node 18 is a breaking change, you need to add the appropriate footer tag for this

@zahin-mohammad When you merge your PR to remove 16, [update the readme too](https://github.com/BitGo/BitGoJS/blob/master/README.md?plain=1#L64C1-L64C7)

zahin-mohammad commented:
> > * Changing min version to node 18 is a breaking change, you need to add the appropriate footer tag for this
> 
> @zahin-mohammad When you merge your PR to remove 16, [update the readme too](https://github.com/BitGo/BitGoJS/blob/master/README.md?plain=1&rgh-link-date=2024-03-26T17%3A35%3A30Z#L64C1-L64C7)

Ah must have missed it, good catch.

zahin-mohammad commented:
> > * Changing min version to node 18 is a breaking change, you need to add the appropriate footer tag for this
> 
> @zahin-mohammad When you merge your PR to remove 16, [update the readme too](https://github.com/BitGo/BitGoJS/blob/master/README.md?plain=1&rgh-link-date=2024-03-26T17%3A35%3A30Z#L64C1-L64C7)

https://github.com/BitGo/BitGoJS/pull/4371/files
